File: templates/dart/examples/protocols_common.dart
export '../lib/mavlink_protocols.dart';
import '../lib/mavlink_protocols.dart';

/// Ground control station identity (MAVLink convention).
const gcsSystemId = 255;
const gcsComponentId = 190;

/// Simulated autopilot identity.
const droneSystemId = 1;
const droneComponentId = 1;

/// Create a linked GCS/drone pair over an in-memory MAVLink bus.
({
  VirtualMavlinkBus bus,
  MavlinkSession gcs,
  MavlinkSession drone,
  MavlinkDialect dialect,
}) createVirtualLink(MavlinkDialect dialect) {
  final bus = VirtualMavlinkBus();
  final gcsLink = bus.createEndpoint();
  final droneLink = bus.createEndpoint();

  final gcs = MavlinkSession(
    dialect: dialect,
    link: gcsLink,
    systemId: gcsSystemId,
    componentId: gcsComponentId,
  );

  final drone = MavlinkSession(
    dialect: dialect,
    link: droneLink,
    systemId: droneSystemId,
    componentId: droneComponentId,
  );

  return (bus: bus, gcs: gcs, drone: drone, dialect: dialect);
}

Future<void> closeVirtualLink({
  required VirtualMavlinkBus bus,
  required MavlinkSession gcs,
  required MavlinkSession drone,
}) async {
  await gcs.close();
  await drone.close();
  await bus.closeAll();
}

File: templates/rust/examples/protocols_common.rs
//! Shared helpers for protocol-class MAVLink examples.

use std::sync::Arc;

pub use mavlink::mavlink_dialect::MavlinkDialect;
pub use mavlink::mavlink_version::MavlinkVersion;
pub use mavlink::protocols::{
    MavlinkLink, MavlinkSession, SessionWaitError, VirtualMavlinkBus,
};

/// Ground control station identity (MAVLink convention).
pub const GCS_SYSTEM_ID: u8 = 255;
pub const GCS_COMPONENT_ID: u8 = 190;

/// Simulated autopilot identity.
pub const DRONE_SYSTEM_ID: u8 = 1;
pub const DRONE_COMPONENT_ID: u8 = 1;

pub struct VirtualLink {
    pub bus: Arc<VirtualMavlinkBus>,
    pub gcs: Arc<MavlinkSession>,
    pub drone: Arc<MavlinkSession>,
    pub dialect: Arc<dyn MavlinkDialect + Send + Sync>,
}

pub fn create_virtual_link(
    dialect: Arc<dyn MavlinkDialect + Send + Sync>,
) -> VirtualLink {
    let bus = VirtualMavlinkBus::new();
    let gcs_link = bus.create_endpoint();
    let drone_link = bus.create_endpoint();

    let gcs = Arc::new(MavlinkSession::new(
        dialect.clone(),
        gcs_link,
        GCS_SYSTEM_ID,
        GCS_COMPONENT_ID,
        MavlinkVersion::V2,
    ));

    let drone = Arc::new(MavlinkSession::new(
        dialect.clone(),
        drone_link,
        DRONE_SYSTEM_ID,
        DRONE_COMPONENT_ID,
        MavlinkVersion::V2,
    ));

    VirtualLink {
        bus,
        gcs,
        drone,
        dialect,
    }
}

pub async fn close_virtual_link(link: VirtualLink) -> Result<(), SessionWaitError> {
    link.gcs.close().await?;
    link.drone.close().await?;
    link.bus.close_all().await;
    Ok(())
}
